feat: integrate BlockchainProvider2 in EthNodeLauncher

fgimenez opened the pull request:
Closes: #9731

There were some references to the concrete type `BlockchainProvider` in the launcher types, this is now generic over `T::Provider`.

Also, in the `with_blockchain_db` method, we need to instantiate the concrete blockchain provider, and the constructors are different for `BlockchainProvider` and `BlockchainProvider2`, the latter doesn't take a `TreeViewer`. Now this is implemented with a closure and the need to create a tree instance outside the ctx methods, once the migration is done this will be simplified.

mattsse reviewed (approved):
cool, this makes sense,
the helper trait will go away eventually, for now this is appropriate

fgimenez commented:
> the helper trait will go away eventually, for now this is appropriate

yep, I've added some comments to keep it in mind https://github.com/paradigmxyz/reth/pull/9754/commits/264f0273d4a9352021a634fc88c25f7c6b3d1f37

> the last thing missing here are type hints for

i've added a call to `with_types_and_provider` specifying the right provider to use, is that enough? https://github.com/paradigmxyz/reth/pull/9754/commits/f9332eabe2494c74c6c3fda53f82bd5684ee31a4

mattsse commented:
> i've added a call to with_types_and_provider specifying the right provider to use, is that enough?

cool, yeah this should do it for now
